Source/WebCore/html/PermissionsPolicy.cpp

```diff
@@ -123,315 +232,148 @@ bool isPermissionsPolicyAllowedByDocumentAndAllOwners(PermissionsPolicy::Type ty
-            else if (auto contentDocument = iframe.contentDocument())
-                allowedOrigin = &contentDocument->securityOrigin();
-            if (allowedOrigin)
-                rule.allowedList.add(allowedOrigin->data());
-        }
-        return;
-    }
+    if (value.isEmpty()) {
+        if (useStarAsDefaultAllowlistValue)
+            return Allowlist { Allowlist::AllowAllOrigins };
 
-    if (item == "*"_s) {
-        rule.type = PermissionsPolicy::AllowRule::Type::All;
-        return;
+        return Allowlist { targetOrigin };
     }
 
-    if (item == "'self'"_s) {
-        rule.allowedList.add(document.securityOrigin().data());
-        return;
-    }
-    if (item == "'none'"_s) {
-        rule.type = PermissionsPolicy::AllowRule::Type::None;
-        return;
+    PermissionsPolicy::Allowlist result;
+    while (!value.isEmpty()) {
+        auto [token, remainingValue] = splitOnAsciiWhiteSpace(value);
+        if (!token.isEmpty()) {
+            if (token == "*"_s)
+                return PermissionsPolicy::Allowlist { Allowlist::AllowAllOrigins };
+
+            if (equalIgnoringASCIICase(token, "'none'"_s))
+                return PermissionsPolicy::Allowlist { };
+
+            if (equalIgnoringASCIICase(token, "'self'"_s))
+                result.addOrigin(containerOrigin);
+            else if (equalIgnoringASCIICase(token, "'src'"_s))
+                result.addOrigin(targetOrigin);
+            else {
+                URL url { { }, token.toString() };
+                if (url.isValid()) {
+                    auto origin = SecurityOriginData::fromURL(url);
+                    if (!origin.isOpaque())
+                        result.addOrigin(origin);
+                }
+            }
+        }
+        value = remainingValue;
     }
-    URL url { { }, item.toString() };
-    if (url.isValid())
-        rule.allowedList.add(SecurityOriginData::fromURL(url));
+
+    return result;
 }
 
-static inline void updateList(Document& document, const HTMLIFrameElement& iframe, PermissionsPolicy::AllowRule& rule, StringView value)
+// https://w3c.github.io/webappsec-permissions-policy/#algo-parse-policy-directive
+PermissionsPolicy::PolicyDirective PermissionsPolicy::parsePolicyDirective(StringView value, const SecurityOriginData& containerOrigin, const SecurityOriginData& targetOrigin, bool useStarAsDefaultAllowlistValue)
 {
-    if (value.isEmpty()) {
-        if (document.quirks().shouldStarBePermissionsPolicyDefaultValue()) {
-            rule.type = PermissionsPolicy::AllowRule::Type::All;
-            return;
-        }
+    PermissionsPolicy::PolicyDirective result;
+    for (auto item : value.split(';')) {
+        // FIXME: Based on spec, we should use splitOnAsciiWhiteSpace to extract feature name,
+        // but some tests would fail as they use ':' as delimiter.
+        auto [feature, remainingItem] = splitFeatureName(item);
+        if (feature == Feature::Invalid)
```

Comment:
I guess we could remove Feature::Invalid and return an optional. Either is fine.